model/pipeline/src/main/resources/org/apache/beam/model/pipeline/pipeline.json

```diff
@@ -0,0 +1,320 @@
+    "UserFn": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // A cross-language, stable, unique identifier for the function. It should indicate
+        // enough information that naive optimization (or language-to-language replacement)
+        // may be possible, since the runner will have no insight into the SDK-specific data.
+        "urn": { "type": "string" },
+
+        // A URL for the container image containing the SDK that understands how to run
+        // this user-definable function.
+        "sdkContainer": { "type": "string" },
+
+        // A free-form object containing SDK-specific data.
+        "data":  { "type": "object" }
+      },
+
+      "required": ["urn", "data"]
+    },
+
+    // A windowing strategy describes the window function, triggering, allowed lateness
+    // for a PCollection
+    "WindowingStrategy": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // The id of a user-definable function that assigns windows, merges windows, and shifts
+        // timestamps.
+        "windowFn": { "type": "string" },
+
+        // Whether or not the window fn is merging; this affects how a runner may implement things
+        //
+        // TBD whether this metadata will live in the UserFn record or here or both
+        "isMergingWindowFn": { "type": "boolean" },
+
+        // The encoding node of the windows for the window_fn.
+        "windowCoder": { "type": "string" },
+
+        // The trigger, as a simple syntax tree
+        // TODO: the referenced definition
+        "trigger": { "$ref": "#/definitions/Trigger" },
+
+        // The accumulation mode indicates whether new panes are a full replacement for
+        // prior panes or whether they are deltas to be combined with other panes (the combine
+        // should correspond to whatever the upstream grouping transform is).
+        "accumulationMode": { "$ref": "#/definitions/AccumulationMode" },
+
+        // The OutputTime specifies, for a grouping transform, how to compute the
+        // aggregate timestamp. The window_fn will first possibly shift it later, then the
+        // OutputTime takes the max, min, or ignores it and takes the end of window.
+        "outputTime": { "$ref": "#/definitions/OutputTime" },
+
+        // After the watermark gets this far past the end of the window, new elements are dropped.
+        // in milliseconds
+        "allowedLateness": { "type": "number" }
+      }
+    },
+
+    "SideInput": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        // Only "multimap" is supported for initial Beam model
+        "accessPattern": { "type": "string" },
+
+        // The SDK converts the access pattern to an SDK-specific type, and this function
+        // converts it to the users desired side input type
+        "viewFn": { "type": "string" },
```

Comment:
Ah, that was unclear, I think by using the "ref" or "id" naming convention it will become clearer.